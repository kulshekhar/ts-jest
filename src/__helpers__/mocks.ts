import { testing } from 'bs-logger'

import { rootLogger } from '../util/logger'

export const logTargetMock = () => (rootLogger as testing.LoggerMock).target

export const mockObject = <T, M>(obj: T, newProps: M): T & M & { mockRestore: () => T } => {
  const backup: any = Object.create(null)

  Object.keys(newProps).forEach(key => {
    const desc = (backup[key] = Object.getOwnPropertyDescriptor(obj, key))
    const newDesc: any = { ...desc }
    if (newDesc.get) {
      newDesc.get = () => (newProps as any)[key]
    } else {
      newDesc.value = (newProps as any)[key]
    }
    Object.defineProperty(obj, key, newDesc)
  })
  if ((obj as any).mockRestore) backup.mockRestore = Object.getOwnPropertyDescriptor(obj, 'mockRestore')
  return Object.defineProperty(obj, 'mockRestore', {
    value() {
      Object.keys(backup).forEach(key => {
        Object.defineProperty(obj, key, backup[key])
      })
      return obj
    },
    configurable: true,
  })
}

export const mockWriteStream = () => ({
  written: [] as string[],
  write(text: string) {
    this.written.push(text)
  },
  clear() {
    this.written = []
  },
})
