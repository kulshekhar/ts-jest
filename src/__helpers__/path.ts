import { ensureDirSync } from 'fs-extra'
import { tmpdir } from 'os'
import { join, resolve } from 'path'

export const ROOT = resolve(__dirname, '..', '..')

export function tempDir(ns: string): string {
  const dir = join(tmpdir(), 'ts-jest-unit-tests', ns, parseInt((Math.random() * 1e17) as any, 10).toString(36))
  ensureDirSync(dir)

  return dir
}
