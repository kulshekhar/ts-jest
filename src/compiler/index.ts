export * from './compiler-utils'
export * from './ts-compiler'
export * from './ts-jest-compiler'
