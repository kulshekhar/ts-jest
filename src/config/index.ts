export * from './config-set'
export * from './paths-to-module-name-mapper'
