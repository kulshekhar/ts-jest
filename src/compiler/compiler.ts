import * as ts from 'typescript';
import * as fs from 'fs';
import * as path from 'path';
import * as utils from './utils';

function formatTscParserErrors(errors: ts.Diagnostic[]) {
  return errors.map(s => JSON.stringify(s, null, 4)).join('\n');
}

interface File {
  path: string;
  version: number;
  snapshot: ts.IScriptSnapshot;
  src: string;
}

class Files {
  private files: Map<string, File>;
  constructor() {
    this.files = new Map();
  }
  // set(path: string, file: File) {
  //     this.files.set(path, file)
  // }
  update({ path, src }: { path: string; src: string }) {
    const file = this.get(path);
    if (file === undefined) {
      const newFile: File = {
        src,
        path,
        version: 0,
        snapshot: ts.ScriptSnapshot.fromString(src),
      };
      this.files.set(path, newFile);
      return;
    }
    if (file.src === src) {
      return;
    }
    file.src = src;
    file.version += 1;
    file.snapshot = ts.ScriptSnapshot.fromString(src);
  }
  get(path: string) {
    return this.files.get(path);
  }
  getFileNames() {
    return Array.from(this.files.values()).map(f => f.path);
  }
  getScriptVersion(path: string) {
    const f = this.get(path);
    if (f === undefined) {
      return '';
    }
    return f.version.toString();
  }
}

export class Compiler {
  private options?: ts.CompilerOptions;
  private service: ts.LanguageService;
  private files: Files;

  constructor(options?: ts.CompilerOptions) {
    this.files = new Files();
    this.options = options;
    this.service = this.createServiceHost();
  }
  setOptions(options: ts.CompilerOptions) {
    this.options = options;
  }
  private createServiceHost() {
    const that = this;

    let service: ts.LanguageService;

    class ServiceHost implements ts.LanguageServiceHost {
      // getCustomTransformers() {
      //     return {
      //         before: [tsruntimeTransformer(service.getProgram())]
      //     }
      // }
      getScriptFileNames() {
        return that.files.getFileNames();
      }
      getScriptVersion(fileName: string) {
        return that.files.getScriptVersion(fileName);
      }
      getScriptSnapshot(fileName: string) {
        const file = that.files.get(fileName);
        if (file !== undefined) {
          return file.snapshot;
        }
        if (!fs.existsSync(fileName)) {
          return undefined;
        }
        return ts.ScriptSnapshot.fromString(
          fs.readFileSync(fileName).toString(),
        ); //todo maybe put to files cache
      }
      getCurrentDirectory = () => process.cwd();
      getCompilationSettings = () => that.options;
      getDefaultLibFileName = (options: ts.CompilerOptions) =>
        ts.getDefaultLibFilePath(options);
      fileExists = ts.sys.fileExists;
      readFile = ts.sys.readFile;
      readDirectory = ts.sys.readDirectory;
      // resolveTypeReferenceDirectives(typeDirectiveNames: string[], containingFile: string) {
      //     const resolved = typeDirectiveNames.map(directive =>
      //         ts.resolveTypeReferenceDirective(directive, containingFile, options, ts.sys)
      //             .resolvedTypeReferenceDirective);

      //     // resolved.forEach(res => {
      //     if (res && res.resolvedFileName) {
      //         fileDeps.add(containingFile, res.resolvedFileName);
      //     }
      // });

      // return resolved;
      // }
    }

    service = ts.createLanguageService(
      new ServiceHost(),
      ts.createDocumentRegistry(),
    );

    return service;
  }

  emitFile({ path, src }: { path: string; src: string }) {
    this.files.update({ path, src });

    let output = this.service.getEmitOutput(path);

    if (output.emitSkipped) {
      const errors = this.formatErrors(path);
      throw new Error(`Errors while compile ts: \n${errors}`);
    }
    const res = utils.findResultFor(path, output);
    return res;
  }

  private formatErrors(fileName: string) {
    let allDiagnostics = this.service
      .getCompilerOptionsDiagnostics()
      .concat(this.service.getSyntacticDiagnostics(fileName))
      .concat(this.service.getSemanticDiagnostics(fileName));

    let res = '';
    allDiagnostics.forEach(diagnostic => {
      let message = ts.flattenDiagnosticMessageText(
        diagnostic.messageText,
        '\n',
      );
      if (diagnostic.file) {
        let { line, character } = diagnostic.file.getLineAndCharacterOfPosition(
          diagnostic.start!,
        );
        res += `  Error ${diagnostic.file.fileName} (${line + 1},${character +
          1}): ${message}\n`;
      } else {
        res += `  Error: ${message}\n`;
      }
    });
    return res;
  }
}
