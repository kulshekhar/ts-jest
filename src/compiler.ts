/**
 * This code is heavilly inspired from
 * https://github.com/JsCommunity/make-error/blob/v1.3.4/index.js
 * ...but more modified than expected :-D
 * Below is the original license anyway:
 *
 * ---
 *
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Blake Embrey (hello@blakeembrey.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import { LogContexts, LogLevels, Logger } from 'bs-logger'
import bufferFrom = require('buffer-from')
import stableStringify = require('fast-json-stable-stringify')
import { readFileSync, writeFileSync } from 'fs'
import memoize = require('lodash.memoize')
import mkdirp = require('mkdirp')
import { basename, extname, join, normalize, relative } from 'path'
import { LanguageServiceHost } from 'typescript'

import { ConfigSet } from './config/config-set'
import { MemoryCache, TsCompiler, TypeInfo } from './types'
import { Errors, interpolate } from './util/messages'
import { sha1 } from './util/sha1'

const hasOwn = Object.prototype.hasOwnProperty

/**
 * Register TypeScript compiler.
 * @internal
 */
export function createCompiler(configs: ConfigSet): TsCompiler {
  const logger = configs.logger.child({ namespace: 'ts-compiler' })
  logger.debug(
    'creating typescript compiler',
    configs.tsJest.isolatedModules ? '(isolated modules)' : '(language service)',
  )
  const cachedir = configs.tsCacheDir

  const memoryCache: MemoryCache = {
    contents: Object.create(null),
    versions: Object.create(null),
    outputs: Object.create(null),
  }

  // Require the TypeScript compiler and configuration.
  const ts = configs.compilerModule
  const cwd = configs.cwd

  const extensions = ['.ts', '.tsx']
  const {
    typescript: { options: compilerOptions, fileNames },
  } = configs

  // Enable `allowJs` when flag is set.
  if (compilerOptions.allowJs) {
    extensions.push('.js')
    extensions.push('.jsx')
  }

  // Initialize files from TypeScript into project.
  for (const path of fileNames) memoryCache.versions[path] = 1

  /**
   * Get the extension for a transpiled file.
   */
  const getExtension =
    compilerOptions.jsx === ts.JsxEmit.Preserve
      ? (path: string) => (/\.[tj]sx$/.test(path) ? '.jsx' : '.js')
      : (_: string) => '.js'

  const transformers = configs.tsCustomTransformers

  /**
   * Create the basic required function using transpile mode.
   */
  let getOutput = (code: string, fileName: string /* , lineOffset = 0 */): SourceOutput => {
    logger.debug({ fileName }, 'getOutput(): compiling as isolated module')
    const result = ts.transpileModule(code, {
      fileName,
      transformers,
      compilerOptions,
      reportDiagnostics: configs.shouldReportDiagnostic(fileName),
    })

    if (result.diagnostics) configs.raiseDiagnostics(result.diagnostics, fileName, logger)

    return [result.outputText, result.sourceMapText as string]
  }

  let getTypeInfo = (_code: string, _fileName: string, _position: number): TypeInfo => {
    throw new TypeError(Errors.TypesUnavailableWithoutTypeCheck)
  }

  // Use full language services when the fast option is disabled.
  if (!configs.tsJest.isolatedModules) {
    // Set the file contents into cache.
    const updateMemoryCache = (code: string, fileName: string) => {
      logger.debug({ fileName }, `updateMemoryCache()`)
      if (memoryCache.contents[fileName] !== code) {
        memoryCache.contents[fileName] = code
        memoryCache.versions[fileName] = (memoryCache.versions[fileName] || 0) + 1
      }
    }

    // Create the compiler host for type checking.
    const serviceHostDebugCtx = {
      [LogContexts.logLevel]: LogLevels.debug,
      namespace: 'ts:serviceHost',
      call: null,
    }
    const serviceHostTraceCtx = {
      ...serviceHostDebugCtx,
      [LogContexts.logLevel]: LogLevels.trace,
    }

    const serviceHost: LanguageServiceHost = {
      getScriptFileNames: () => Object.keys(memoryCache.versions),
      getScriptVersion: (fileName: string) => {
        const normalizedFileName = normalize(fileName)
        const version = memoryCache.versions[normalizedFileName]

        // We need to return `undefined` and not a string here because TypeScript will use
        // `getScriptVersion` and compare against their own version - which can be `undefined`.
        // If we don't return `undefined` it results in `undefined === "undefined"` and run
        // `createProgram` again (which is very slow). Using a `string` assertion here to avoid
        // TypeScript errors from the function signature (expects `(x: string) => string`).
        return version === undefined ? ((undefined as any) as string) : String(version)
      },
      getScriptSnapshot(fileName: string) {
        const normalizedFileName = normalize(fileName)
        const hit = hasOwn.call(memoryCache.contents, normalizedFileName)
        logger.trace({ normalizedFileName, cacheHit: hit }, `getScriptSnapshot():`, 'cache', hit ? 'hit' : 'miss')
        // Read contents from TypeScript memory cache.
        if (!hit) {
          memoryCache.contents[normalizedFileName] = ts.sys.readFile(normalizedFileName)
        }

        const contents = memoryCache.contents[normalizedFileName]
        if (contents === undefined) {
          return
        }
        return ts.ScriptSnapshot.fromString(contents)
      },
      fileExists: memoize(ts.sys.fileExists),
      readFile: logger.wrap(serviceHostTraceCtx, 'readFile', memoize(ts.sys.readFile)),
      readDirectory: memoize(ts.sys.readDirectory),
      getDirectories: memoize(ts.sys.getDirectories),
      directoryExists: memoize(ts.sys.directoryExists),
      realpath: memoize(ts.sys.realpath!),
      getNewLine: () => '\n',
      getCurrentDirectory: () => cwd,
      getCompilationSettings: () => compilerOptions,
      getDefaultLibFileName: () => ts.getDefaultLibFilePath(compilerOptions),
      getCustomTransformers: () => transformers,
    }

    logger.debug('creating language service')
    const service = ts.createLanguageService(serviceHost)

    getOutput = (code: string, fileName: string /*, lineOffset = 0 */) => {
      logger.debug({ fileName }, 'getOutput(): compiling using language service')
      // Must set memory cache before attempting to read file.
      updateMemoryCache(code, fileName)

      const output = service.getEmitOutput(fileName)

      if (configs.shouldReportDiagnostic(fileName)) {
        logger.debug({ fileName }, 'getOutput(): computing diagnostics')
        // Get the relevant diagnostics - this is 3x faster than `getPreEmitDiagnostics`.
        const diagnostics = service
          .getCompilerOptionsDiagnostics()
          .concat(service.getSyntacticDiagnostics(fileName))
          .concat(service.getSemanticDiagnostics(fileName))

        // will raise or just warn diagnostics depending on config
        configs.raiseDiagnostics(diagnostics, fileName, logger)
      }

      /* istanbul ignore next (this should never happen but is kept for security) */
      if (output.emitSkipped) {
        throw new TypeError(`${relative(cwd, fileName)}: Emit skipped`)
      }

      // Throw an error when requiring `.d.ts` files.
      /* istanbul ignore next (this should never happen but is kept for security) */
      if (output.outputFiles.length === 0) {
        throw new TypeError(
          interpolate(Errors.UnableToRequireDefinitionFile, {
            file: basename(fileName),
          }),
        )
      }

      return [output.outputFiles[1].text, output.outputFiles[0].text]
    }

    getTypeInfo = (code: string, fileName: string, position: number) => {
      updateMemoryCache(code, fileName)

      const info = service.getQuickInfoAtPosition(fileName, position)
      const name = ts.displayPartsToString(info ? info.displayParts : [])
      const comment = ts.displayPartsToString(info ? info.documentation : [])

      return { name, comment }
    }
  }

  const compile = readThrough(cachedir, memoryCache, getOutput, getExtension, cwd, logger)
  return { cwd, compile, getTypeInfo, extensions, cachedir, ts }
}

/**
 * Internal source output.
 */
type SourceOutput = [string, string]

/**
 * Wrap the function with caching.
 */
function readThrough(
  cachedir: string | undefined,
  memoryCache: MemoryCache,
  compile: (code: string, fileName: string, lineOffset?: number) => SourceOutput,
  getExtension: (fileName: string) => string,
  cwd: string,
  logger: Logger,
) {
  if (!cachedir) {
    return (code: string, fileName: string, lineOffset?: number) => {
      logger.debug({ fileName }, 'readThrough(): no cache')

      const [value, sourceMap] = compile(code, fileName, lineOffset)
      const output = updateOutput(value, fileName, sourceMap, getExtension, cwd)

      memoryCache.outputs[fileName] = output

      return output
    }
  }

  // Make sure the cache directory exists before continuing.
  mkdirp.sync(cachedir)

  return (code: string, fileName: string, lineOffset?: number) => {
    const cachePath = join(cachedir, getCacheName(code, fileName))
    const extension = getExtension(fileName)
    const outputPath = `${cachePath}${extension}`

    try {
      const output = readFileSync(outputPath, 'utf8')
      if (isValidCacheContent(output)) {
        logger.debug({ fileName }, 'readThrough(): cache hit')
        memoryCache.outputs[fileName] = output
        return output
      }
    } catch (err) {}

    logger.debug({ fileName }, 'readThrough(): cache miss')
    const [value, sourceMap] = compile(code, fileName, lineOffset)
    const output = updateOutput(value, fileName, sourceMap, getExtension, cwd)

    logger.debug({ fileName, outputPath }, 'readThrough(): writing caches')
    memoryCache.outputs[fileName] = output
    writeFileSync(outputPath, output)

    return output
  }
}

/**
 * Update the output remapping the source map.
 */
function updateOutput(
  outputText: string,
  fileName: string,
  sourceMap: string,
  getExtension: (fileName: string) => string,
  sourceRoot: string,
) {
  const base = basename(fileName)
  const base64Map = bufferFrom(updateSourceMap(sourceMap, fileName, sourceRoot), 'utf8').toString('base64')
  const sourceMapContent = `data:application/json;charset=utf-8;base64,${base64Map}`
  const sourceMapLength = `${base}.map`.length + (getExtension(fileName).length - extname(fileName).length)

  return outputText.slice(0, -sourceMapLength) + sourceMapContent
}

/**
 * Update the source map contents for improved output.
 */
function updateSourceMap(sourceMapText: string, fileName: string, _sourceRoot: string) {
  const sourceMap = JSON.parse(sourceMapText)
  // const relativeFilePath = posix.normalize(relative(sourceRoot, fileName))
  // sourceMap.file = relativeFilePath
  // sourceMap.sources = [relativeFilePath]
  // sourceMap.sourceRoot = normalize(sourceRoot)
  sourceMap.file = fileName
  sourceMap.sources = [fileName]
  delete sourceMap.sourceRoot
  return stableStringify(sourceMap)
}

/**
 * Get the file name for the cache entry.
 */
function getCacheName(sourceCode: string, fileName: string) {
  return sha1(fileName, '\x00', sourceCode)
}

/**
 * Ensure the given cached content is valid by sniffing for a base64 encoded '}'
 * at the end of the content, which should exist if there is a valid sourceMap present.
 */
function isValidCacheContent(contents: string) {
  return /(?:9|0=|Q==)$/.test(contents.slice(-3))
}
