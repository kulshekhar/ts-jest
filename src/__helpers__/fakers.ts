import { Config } from '@jest/types'
import { resolve } from 'path'

import { createCompilerInstance } from '../compiler/instance'
import { ConfigSet } from '../config/config-set'
import { BabelConfig, TsCompiler, TsJestConfig, TsJestGlobalOptions } from '../types'
import { ImportReasons } from '../util/messages'

export function filePath(relPath: string): string {
  return resolve(__dirname, '..', '..', relPath)
}

export const rootDir = filePath('')

export function tsJestConfig(options?: Partial<TsJestConfig>): TsJestConfig {
  return {
    isolatedModules: false,
    compiler: 'typescript',
    transformers: [],
    babelConfig: undefined,
    tsConfig: undefined,
    packageJson: undefined,
    stringifyContentPathRegex: undefined,
    diagnostics: { ignoreCodes: [], pretty: false, throws: true },
    ...options,
  }
}

function getJestConfig<T extends Config.ProjectConfig>(
  options?: Partial<Config.InitialOptions | Config.ProjectConfig>,
  tsJestOptions?: TsJestGlobalOptions,
): T {
  const res = {
    globals: {},
    ...options,
  } as any
  if (tsJestOptions) {
    res.globals['ts-jest'] = tsJestOptions
  }

  return res
}

export function babelConfig<T extends BabelConfig>(options?: BabelConfig): T {
  return {
    ...options,
    presets: [...(options && options.presets)],
    plugins: [...(options && options.plugins)],
  } as any
}

export function importReason(text = '[[BECAUSE]]'): ImportReasons {
  return text as any
}

export const defaultResolve = (path: string): string => `resolved:${path}`

export function createConfigSet({
  jestConfig,
  tsJestConfig,
  parentConfig,
  resolve = defaultResolve,
  ...others
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
  resolve?: ((path: string) => string) | null
  [key: string]: any
} = {}): ConfigSet {
  const defaultTestRegex = ['(/__tests__/.*|(\\\\.|/)(test|spec))\\\\.[jt]sx?$']
  const defaultTestMatch = ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)']
  jestConfig = {
    ...jestConfig,
    testMatch: jestConfig?.testMatch ? [...jestConfig.testMatch, ...defaultTestMatch] : defaultTestMatch,
    testRegex: jestConfig?.testRegex ? [...defaultTestRegex, ...jestConfig.testRegex] : defaultTestRegex,
  }
  const cs = new ConfigSet(getJestConfig(jestConfig, tsJestConfig), parentConfig)
  if (resolve) {
    cs.resolvePath = resolve
  }
  Object.keys(others).forEach((key) => {
    Object.defineProperty(cs, key, { value: others[key] })
  })

  return cs
}

// not really unit-testing here, but it's hard to mock all those values :-D
export function makeCompiler({
  jestConfig,
  tsJestConfig,
  parentConfig,
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
} = {}): TsCompiler {
  tsJestConfig = { ...tsJestConfig }
  tsJestConfig.diagnostics = {
    ...(tsJestConfig.diagnostics as any),
    pretty: false,
  }
  const cs = createConfigSet({ jestConfig, tsJestConfig, parentConfig, resolve: null })

  return createCompilerInstance(cs)
}
