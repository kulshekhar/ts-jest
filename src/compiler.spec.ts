// tslint:disable:max-line-length
import { Config } from '@jest/types'
import { LogLevels } from 'bs-logger'
import { removeSync, writeFileSync } from 'fs-extra'
import { LanguageService, ModuleKind, ScriptTarget } from 'typescript'

import * as fakers from './__helpers__/fakers'
import { logTargetMock } from './__helpers__/mocks'
import { tempDir } from './__helpers__/path'
import ProcessedSource from './__helpers__/processed-source'
import { createCompiler } from './compiler'
import { ConfigSet } from './config/config-set'
import { TsJestGlobalOptions } from './types'

const logTarget = logTargetMock()

// not really unit-testing here, but it's hard to mock all those values :-D
function makeCompiler({
  jestConfig,
  tsJestConfig,
  parentConfig,
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
} = {}) {
  tsJestConfig = { ...tsJestConfig }
  tsJestConfig.diagnostics = {
    ...(tsJestConfig.diagnostics as any),
    pretty: false,
  }
  const cs = new ConfigSet(fakers.jestConfig(jestConfig, tsJestConfig), parentConfig)

  return createCompiler(cs)
}

function makeFallbackCompiler({
  jestConfig,
  tsJestConfig,
  parentConfig,
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
} = {}) {
  const compiler = makeCompiler({ jestConfig, tsJestConfig, parentConfig })
  const languageService: LanguageService = compiler.ts.createLanguageService({
    getScriptFileNames: jest.fn(),
    getScriptVersion: jest.fn(),
    getScriptSnapshot: jest.fn(),
    getCurrentDirectory: () => '.',
    getDefaultLibFileName: jest.fn(),
    getCompilationSettings: () => {
      return {
        target: ScriptTarget.ES2018,
        module: ModuleKind.CommonJS,
        lib: ['dom', 'es2018'],
      }
    },
  })
  compiler.ts.createLanguageService = jest.fn().mockReturnValue(languageService)
  languageService.getEmitOutput = jest.fn().mockReturnValue({
    outputFiles: [],
    emitSkipped: false,
  })

  return compiler
}

beforeEach(() => {
  logTarget.clear()
})

describe('typings', () => {
  const compiler = makeCompiler({ tsJestConfig: { tsConfig: false } })
  it('should report diagnostics related to typings', () => {
    expect(() =>
      compiler.compile(
        `
const f = (v: number) => v
const t: string = f(5)
const v: boolean = t
`,
        'foo.ts',
      ),
    ).toThrowErrorMatchingInlineSnapshot(`
"TypeScript diagnostics (customize using \`[jest-config].globals.ts-jest.diagnostics\` option):
foo.ts(3,7): error TS2322: Type 'number' is not assignable to type 'string'.
foo.ts(4,7): error TS2322: Type 'string' is not assignable to type 'boolean'."
`)
  })
})

describe('source-maps', () => {
  const compiler = makeCompiler({ tsJestConfig: { tsConfig: false } })
  it('should have correct source maps', () => {
    const source = 'const f = (v: number) => v\nconst t: number = f(5)'
    const compiled = compiler.compile(source, __filename)
    const processed = new ProcessedSource(compiled, __filename)
    // const expectedFileName = relativeToRoot(__filename)
    const expectedFileName = __filename
    expect(processed.outputSourceMaps).toMatchObject({
      file: expectedFileName,
      sources: [expectedFileName],
      sourcesContent: [source],
    })
  })
})

describe('cache', () => {
  const tmp = tempDir('compiler')
  const compiler = makeCompiler({
    jestConfig: { cache: true, cacheDirectory: tmp },
    tsJestConfig: { tsConfig: false },
  })
  const source = 'console.log("hello")'

  it('should use the cache', () => {
    const compiled1 = compiler.compile(source, __filename)
    expect(logTarget.filteredLines(LogLevels.debug, Infinity)).toMatchInlineSnapshot(`
      Array [
        "[level:20] readThrough(): cache miss
      ",
        "[level:20] getOutput(): compiling using language service
      ",
        "[level:20] updateMemoryCache()
      ",
        "[level:20] visitSourceFileNode(): hoisting
      ",
        "[level:20] getOutput(): computing diagnostics from language service
      ",
        "[level:20] readThrough(): writing caches
      ",
      ]
    `)

    logTarget.clear()
    const compiled2 = compiler.compile(source, __filename)
    expect(logTarget.lines).toMatchInlineSnapshot(`
Array [
  "[level:20] readThrough(): cache hit
",
]
`)

    expect(new ProcessedSource(compiled1, __filename)).toMatchSnapshot()
    expect(compiled2).toBe(compiled1)
  })
})

describe('isolatedModules', () => {
  const compiler = makeCompiler({ tsJestConfig: { isolatedModules: true, tsConfig: false } })
  const spy = jest.spyOn(require('typescript'), 'transpileModule')
  afterAll(() => {
    spy.mockRestore()
  })
  it('should compile using transpileModule', () => {
    const compiled = compiler.compile('export default 42', __filename)
    expect(new ProcessedSource(compiled, __filename)).toMatchSnapshot()
    expect(spy).toHaveBeenCalled()
  })
})

describe('allowJs', () => {
  const compiler = makeCompiler({ tsJestConfig: { tsConfig: { allowJs: true } } })
  const fileName = `${__filename}.test.js`
  afterAll(() => {
    removeSync(fileName)
  })
  it('should compile js file', () => {
    const source = 'export default 42'
    writeFileSync(fileName, source, 'utf8')
    const compiled = compiler.compile(source, fileName)
    const processed = new ProcessedSource(compiled, fileName)
    expect(processed).toMatchSnapshot()
  })
})

describe('getTypeInfo', () => {
  const compiler = makeCompiler({ tsJestConfig: { tsConfig: false } })
  const source = `
type MyType {
  /** the prop 1! */
  p1: boolean
}
const val: MyType = {} as any
console.log(val.p1/* <== that */)
`
  it('should get correct type info', () => {
    const ti = compiler.getTypeInfo(source, __filename, source.indexOf('/* <== that */') - 1)
    // before TS 3.1 the comment had an extra tailing space
    ti.comment = ti.comment.trim()
    expect(ti).toEqual({
      comment: 'the prop 1!',
      name: '(property) p1: boolean',
    })
  })
})

describe('Program', () => {
  // These preparation steps are needed to make the test work correctly
  const tmp = tempDir('compiler')
  let compiler = makeFallbackCompiler({
    jestConfig: { cache: true, cacheDirectory: tmp },
    tsJestConfig: { tsConfig: false },
  })
  const source = 'console.log("hello")'
  it('should compile using Program as fallback', () => {
    try {
      compiler = makeFallbackCompiler({ tsJestConfig: { tsConfig: false } })
      compiler.compile(source, 'test.ts')
    } catch (e) {
      expect(e.message).toMatchSnapshot('compile-error')
      expect(logTarget.filteredLines(LogLevels.debug, Infinity)).toMatchInlineSnapshot(`
        Array [
          "[level:20] backporting config
        ",
          "[level:20] normalized jest config
        ",
          "[level:20] normalized ts-jest config
        ",
          "[level:20] creating typescript compiler (language service)
        ",
          "[level:20] file caching disabled
        ",
          "[level:20] normalized typescript config
        ",
          "[level:20] creating language service
        ",
          "[level:20] readThrough(): no cache
        ",
          "[level:20] getOutput(): compiling using language service
        ",
          "[level:20] updateMemoryCache()
        ",
          "[level:20] getOutput(): creating Program as fallback for language service
        ",
          "[level:20] getOutput(): compiling using Program
        ",
          "[level:20] getOutput(): computing diagnostics from Program emit result
        ",
        ]
      `)
    }
  })
})
