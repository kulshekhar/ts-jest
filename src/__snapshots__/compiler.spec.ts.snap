// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Program should compile using Program as fallback: compile-error 1`] = `
"Unable to require \`.d.ts\` file.
This is usually the result of a faulty configuration or import. Make sure there is a \`.js\`, \`.json\` or another executable extension available alongside \`test.ts\`."
`;

exports[`allowJs should compile js file 1`] = `
  ===[ FILE: src/compiler.spec.ts.test.js ]=======================================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.default = 42;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiPGN3ZD4vc3JjL2NvbXBpbGVyLnNwZWMudHMudGVzdC5qcyIsIm1hcHBpbmdzIjoiOztBQUFBLGtCQUFlLEVBQUUsQ0FBQSIsIm5hbWVzIjpbXSwic291cmNlcyI6WyI8Y3dkPi9zcmMvY29tcGlsZXIuc3BlYy50cy50ZXN0LmpzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBkZWZhdWx0IDQyIl0sInZlcnNpb24iOjN9
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: <cwd>/src/compiler.spec.ts.test.js
  mappings: ';;AAAA,kBAAe,EAAE,CAAA'
  names: []
  sources:
    - <cwd>/src/compiler.spec.ts.test.js
  sourcesContent:
    - export default 42
  version: 3
  ================================================================================
`;

exports[`cache should use the cache 3`] = `
  ===[ FILE: src/compiler.spec.ts ]===============================================
  console.log("hello");
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiPGN3ZD4vc3JjL2NvbXBpbGVyLnNwZWMudHMiLCJtYXBwaW5ncyI6IkFBQUEsT0FBTyxDQUFDLEdBQUcsQ0FBQyxPQUFPLENBQUMsQ0FBQSIsIm5hbWVzIjpbXSwic291cmNlcyI6WyI8Y3dkPi9zcmMvY29tcGlsZXIuc3BlYy50cyJdLCJzb3VyY2VzQ29udGVudCI6WyJjb25zb2xlLmxvZyhcImhlbGxvXCIpIl0sInZlcnNpb24iOjN9
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: <cwd>/src/compiler.spec.ts
  mappings: 'AAAA,OAAO,CAAC,GAAG,CAAC,OAAO,CAAC,CAAA'
  names: []
  sources:
    - <cwd>/src/compiler.spec.ts
  sourcesContent:
    - console.log("hello")
  version: 3
  ================================================================================
`;

exports[`isolatedModules should compile using transpileModule 1`] = `
  ===[ FILE: src/compiler.spec.ts ]===============================================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.default = 42;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiPGN3ZD4vc3JjL2NvbXBpbGVyLnNwZWMudHMiLCJtYXBwaW5ncyI6Ijs7QUFBQSxrQkFBZSxFQUFFLENBQUEiLCJuYW1lcyI6W10sInNvdXJjZXMiOlsiPGN3ZD4vc3JjL2NvbXBpbGVyLnNwZWMudHMiXSwic291cmNlc0NvbnRlbnQiOlsiZXhwb3J0IGRlZmF1bHQgNDIiXSwidmVyc2lvbiI6M30=
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: <cwd>/src/compiler.spec.ts
  mappings: ';;AAAA,kBAAe,EAAE,CAAA'
  names: []
  sources:
    - <cwd>/src/compiler.spec.ts
  sourcesContent:
    - export default 42
  version: 3
  ================================================================================
`;
